### DISCUSSION TITLE
Detectron2 PointRend on prediction not giving masks

### DISCUSSION THREAD

**Discussion by hiroto01** [facebookresearch/detectron2#4146] [Q&A]
my code:

visualizing and setting configs: 

    dataset_dicts_poly = my_dataset_function("train")
    for d in random.sample(dataset_dicts_poly, 3):
        # print(d)
        img = cv2.imread(d['file_name'], cv2.IMREAD_UNCHANGED)
        print(d["file_name"])
        if img.shape[2] == 4:     # we have an alpha channel
              # print("alpha channel")
              a1 = ~img[:,:,3]        # extract and invert that alpha
              img = cv2.add(cv2.merge([a1,a1,a1,a1]), img)   # add up values (with clipping)
              img = cv2.cvtColor(img, cv2.COLOR_RGBA2RGB)
        plt.imshow(img[:, :, ::-1])
        plt.show()
        visualizer = Visualizer(img, metadata=stroke_metadata, scale=1.0)
        # print(d)
        out = visualizer.draw_dataset_dict(d)
        # print(out)
        plt.imshow((out.get_image()[:, :, ::-1]))
    # plt.imshow(cv2.cvtColor(out.get_image()[:, :, ::-1], cv2.COLOR_BGR2RGB))
    plt.show()

        from detectron2.engine import DefaultTrainer
    from detectron2.config import get_cfg
    import detectron2.data.transforms as T
    from detectron2.projects.point_rend import add_pointrend_config
    
    cfg = get_cfg()
    point_rend.add_pointrend_config(cfg)
    cfg.merge_from_file("/detectron2_repo/projects/PointRend/configs/InstanceSegmentation/pointrend_rcnn_R_50_FPN_3x_coco.yaml")
    cfg.DATASETS.TRAIN = ("dataSet11_train",)
    cfg.DATASETS.TEST = ()
    
    cfg.DATALOADER.NUM_WORKERS = 0
    cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.7
    cfg.MODEL.WEIGHTS = "detectron2://PointRend/InstanceSegmentation/pointrend_rcnn_R_50_FPN_3x_coco/164955410/model_final_edd263.pkl"
    cfg.SOLVER.IMS_PER_BATCH = 2
    cfg.SOLVER.BASE_LR = 0.00025
    cfg.SOLVER.MAX_ITER = 500
    cfg.SOLVER.STEPS = []        # do not decay learning rate
    # cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128
    cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1
    cfg.MODEL.POINT_HEAD.NUM_CLASSES = 1
    cfg.MODEL.DEVICE = "cpu"
    
    
    
    
    cfg.OUTPUT_DIR = "/path"
    os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
    
    trainer = DefaultTrainer(cfg) 
    trainer.resume_or_load(resume=False)
    trainer.train()
```


now instance prediction:

  ```
  cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
    cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.5
    cfg.DATASETS.TEST = ("dataSet21_test_sample", )
    predictor = DefaultPredictor(cfg)
    
    from detectron2.utils.visualizer import ColorMode
    dataset_dicts = my_dataset_function("test")
    
    
    for d in random.sample(dataset_dicts, 5):    
        im = cv2.imread(d["file_name"], cv2.IMREAD_UNCHANGED)
        print(cwd)
        print(d["file_name"])
        # print(im)
        # height, width = im.shape[:2]
        print(cwd)
        if im.shape[2] == 4:     # we have an alpha channel
          a1 = ~im[:,:,3]        # extract and invert that alpha
          im = cv2.add(cv2.merge([a1,a1,a1,a1]), im)   # add up values (with clipping)
          im = cv2.cvtColor(im, cv2.COLOR_RGBA2RGB)  
        
        outputs = predictor(im)
        print(outputs["instances"].pred_classes)
        print(outputs["instances"].pred_boxes)
        
        cv2_imshow(im[:, :, ::-1])
        # plt.imshow(im[:, :, ::-1])
        # plt.show()
        outputs = predictor(im)
        v = Visualizer(im[:, :, ::-1],
                    metadata=stroke_metadata, 
                    scale=1.0
        )
        v = v.draw_instance_predictions(outputs["instances"].to("cpu"))
        plt.imshow(cv2.cvtColor(v.get_image()[:, :, ::-1], cv2.COLOR_BGR2RGB))
        plt.show()
```


in the output i get no masks and the bounding boxes arent accurate and are so many!!!

outputs for :
outputs = predictor(im)
print(outputs)

is:

    {'instances': Instances(num_instances=17, image_height=249, image_width=256, fields=[pred_boxes: Boxes(tensor([[111.3855,  44.9594, 201.7305, 174.3821],
            [  0.9401,  89.5763,  93.9409, 192.7719],
            [  0.4979,   0.0000,  72.5964,  72.4459],
            [  0.0000,  39.4311, 174.0473, 165.0736],
            [  0.0000,  47.4863, 111.6758, 238.0947],
            [100.0813,  58.2505, 215.4891, 130.5227],
            [112.5025,   0.0000, 246.5010, 156.3627],
            [124.5675,   0.0000, 191.4291, 146.4501],
            [ 26.4513,  95.5350, 157.1500, 176.4264],
            [160.9344,   0.0000, 200.0611,  52.6539],
            [158.7462,   0.0000, 212.2953,  97.9019],
            [118.5634,  91.2083, 233.4940, 154.9819],
            [  0.0000,   0.0000, 114.4604, 101.4680],
            [  0.0000,   0.0000, 121.5726,  42.5086],
            [149.7184,   2.6170, 247.9497,  73.9035],
            [ 92.2937, 103.3998, 221.9623, 192.5935],
            [176.2557,   0.0000, 233.3404, 109.8043]])), scores: tensor([0.5092, 0.4984, 0.4284, 0.3980, 0.3717, 0.3456, 0.3130, 0.2852, 0.2805,
            0.2705, 0.2668, 0.2604, 0.2376, 0.2297, 0.2184, 0.2125, 0.2041]), pred_classes: tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), pred_masks: tensor([[[False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             ...,
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False]],
    
            [[False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             ...,
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False]],
    
            [[False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             ...,
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False]],
    
            ...,
    
            [[False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             ...,
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False]],
    
            [[False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             ...,
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False]],
    
            [[False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             ...,
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False],
             [False, False, False,  ..., False, False, False]]])])}


i have made use of detectron2's pointRend for instance segmentation where i need to  find the strokes over the image(i send image masks along) . on training it aachieves good accuracy but then, on prediction th eoutput is not what i desire. Unlike where in visualizing before the training step it showed the correct output. any help would be highly appreciated.

**Comment by hiroto01** [comment]
@ppwwyyxx sorry to tag you but can you please help with this?

**Comment by hiroto01** [comment]
any help is appreciated. thanks
